test/bpf_tests/bpf_test.go

```diff
@@ -0,0 +1,470 @@
+			},
+		})
+		if err != nil {
+			// Use this definition of ENOSPC instead of syscall.ENOSPC, since the stdlib constant is only
+			// available on linux. Even if this doesn't run, we still want it to compile on non-linux systems.
+			const ENOSPC = syscall.Errno(0x1c)
+			// ENOSPC usually means that the log size was to small, so increase and retry.
+			if errors.Is(err, ENOSPC) {
+				// 1, 4, 16, 64, 256 MiB
+				logSize = logSize << 2
+				continue
+			}
+
+			t.Fatal("new coll:", err)
+		}
+
+		break
+	}
+	defer coll.Close()
+
+	err = loadCallsMap(spec, coll)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testNameToPrograms := make(map[string]programSet)
+
+	for progName, spec := range spec.Programs {
+		match := checkProgRegex.FindStringSubmatch(spec.SectionName)
+		if len(match) == 0 {
+			continue
+		}
+
+		progs := testNameToPrograms[match[1]]
+		if match[2] == "setup" {
+			progs.setupProg = coll.Programs[progName]
+		}
+		if match[2] == "check" {
+			progs.checkProg = coll.Programs[progName]
+		}
+		testNameToPrograms[match[1]] = progs
+	}
+
+	for progName, set := range testNameToPrograms {
+		if set.checkProg == nil {
+			t.Fatalf(
+				"File '%s' contains a setup program in section '%s' but no check program.",
+				elfPath,
+				spec.Programs[progName].SectionName,
+			)
+		}
+	}
+
+	// Collect debug events and add them as logs of the main test
+	var globalLogReader *perf.Reader
+	if coll.Maps["test_cilium_events"] != nil {
+		globalLogReader, err = perf.NewReader(coll.Maps["test_cilium_events"], 4096*16)
+		if err != nil {
+			t.Fatalf("new global log reader: %s", err.Error())
+		}
+		defer globalLogReader.Close()
+
+		linkCache := link.NewLinkCache()
+
+		go func() {
+			for {
+				rec, err := globalLogReader.Read()
+				if err != nil {
+					return
+				}
+
+				dm := monitor.DebugMsg{}
+				reader := bytes.NewReader(rec.RawSample)
+				if err := binary.Read(reader, byteorder.Native, &dm); err != nil {
+					return
+				}
+
+				t.Log(dm.Message(linkCache))
+			}
+		}()
+	}
+
+	for name, progs := range testNameToPrograms {
+		t.Run(name, subTest(progs, coll.Maps[suiteResultMap]))
+	}
+
+	if globalLogReader != nil {
+		// Give the global log buf some time to empty
+		// TODO: replace with flush on the buffer, as soon as cilium/ebpf supports that
+		time.Sleep(50 * time.Millisecond)
+	}
+}
+
+func loadAndPrepSpec(t *testing.T, elfPath string) *ebpf.CollectionSpec {
+	spec, err := ebpf.LoadCollectionSpec(elfPath)
+	if err != nil {
+		t.Fatal("load spec: ", elfPath, err)
+	}
+
+	for _, mspec := range spec.Maps {
+		// Drain extra info from map specs, cilium/ebpf will complain if we don't
+		if mspec.Extra != nil {
+			io.ReadAll(mspec.Extra)
+		}
+
+		// Remove BTF from maps as a workaround since some maps, like 'cilium_xdp_scratch' will return a
+		// "map create: invalid argument" error if we don't
+		mspec.BTF = nil
+	}
+
+	// Detect program type mismatches
+	var progTestType ebpf.ProgramType
+	for _, spec := range spec.Programs {
```

Comment:
This shadowing of `spec` is rather confusing:
```suggestion
	for _, prog := range spec.Programs {
```